vk/response.go

```diff
@@ -1,77 +1,72 @@
 
-// TODO: add convenience helpers for status codes
-
-const (
-	contentTypeJSON        contentType = "application/json"
-	contentTypeTextPlain   contentType = "text/plain"
-	contentTypeOctetStream contentType = "application/octet-stream"
-)
-
-// converts _something_ into bytes, best it can:
-// if data is Response type, returns (status, body processed as below)
-// if bytes, return (200, bytes)
-// if string, return (200, []byte(string))
-// if struct, return (200, json(struct))
-// otherwise, return (500, nil)
-func responseOrOtherToBytes(l *vlog.Logger, data interface{}) (int, []byte, contentType) {
-	if data == nil {
-		return http.StatusNoContent, []byte{}, contentTypeTextPlain
+// RespondString sends the data as a raw string to the client. Ctx is a placeholder here, it is currently unused, but
+// will be used for tracing / logging help purposes later.
+func RespondString(ctx context.Context, w http.ResponseWriter, data string, statusCode int) error {
+	// If there is nothing to marshal then set status code and return.
+	if statusCode == http.StatusNoContent {
+		w.WriteHeader(statusCode)
+		return nil
 	}
 
-	statusCode := http.StatusOK
-	realData := data
+	w.Header().Set("Content-Type", "text/plain")
 
-	// first, check if it's response type, and unpack it for further processing
-	if r, ok := data.(Response); ok {
-		statusCode = r.status
-		realData = r.body
-	}
+	return respondBytes(ctx, w, []byte(data), statusCode)
+}
 
-	// if data is []byte or string, return it as-is
-	if b, ok := realData.([]byte); ok {
-		return statusCode, b, contentTypeOctetStream
-	} else if s, ok := realData.(string); ok {
-		return statusCode, []byte(s), contentTypeTextPlain
+// RespondBytes takes the content we want to send back to the client as byte slice, does an early exit for no content,
+// and then straight pipes into the private respondBytes. This is in front of the private func because of pattern
+// consistency and making sure that a 204 with not empty content does not get written to the ResponseWriter.
+func RespondBytes(ctx context.Context, w http.ResponseWriter, data []byte, statusCode int) error {
+	// If there is nothing to marshal then set status code and return.
+	if statusCode == http.StatusNoContent {
+		w.WriteHeader(statusCode)
+		return nil
 	}
 
-	// otherwise, assume it's a struct of some kind,
-	// so JSON marshal it and return it
-	json, err := json.Marshal(realData)
-	if err != nil {
-		l.Error(errors.Wrap(err, "failed to Marshal response struct"))
+	w.Header().Set("Content-Type", "text/plain")
+
+	return respondBytes(ctx, w, data, statusCode)
+}
+
+// respondBytes takes content as []byte and pipes it into the http.ResponseWriter. The assumption is that the content
+// type header has already been set. The only time this is not called is when the status code is http.StatusNoContent,
+// which would have broken early in the callers.
+func respondBytes(_ context.Context, w http.ResponseWriter, content []byte, statusCode int) error {
```

Comment:
Should content type be an argument here so we don't need to repeat `w.Header().Set("Content-Type",  {type}) in each response formatter?

Reply:
My aim here is that whatever we're returning as data, the data itself needs to be passed in bytes, however the shape of the data, ie setting the mime type, should be decided by the individual handlers.

If we set the content type here (raw bytes), then we already decided for the handler functions what the client is going to view their data as, which adds a lot of friction. Not setting the content type here is a conscious choice I made.